image_builder/image_builder/git.py

```diff
@@ -1,41 +1,40 @@
 from pathlib import Path
+import shutil
 from typing import Optional
 from urllib.parse import urlsplit, quote
 
-from git import Repo
-
 from racetrack_commons.dir import project_root
 from racetrack_client.log.context_error import wrap_context, ContextError
 from racetrack_client.client_config.client_config import Credentials
 from racetrack_client.manifest import Manifest
 from racetrack_client.utils.shell import shell, CommandError, shell_output
 
 
-def fetch_repository(workspace: Path, manifest: Manifest, git_credentials: Optional[Credentials]) -> Path:
+def fetch_repository(workspace: Path, manifest: Manifest, git_credentials: Optional[Credentials]) -> tuple[Path, str]:
     """Clone git repository to local workspace directory and return fetched project path"""
     remote_url = git_remote_with_credentials(manifest.git.remote, git_credentials)
 
     with wrap_context('cloning git repo'):
         try:
-            shell(f'GIT_TERMINAL_PROMPT=0 git clone {remote_url} {workspace.resolve()}',
-                  workdir=project_root(), print_log=False)
+            if manifest.git.branch:
+                assert '"' not in manifest.git.branch, 'illegal characters in a branch name'
+            branch_cmd = f'--branch "{manifest.git.branch}"' if manifest.git.branch else ''
+            cmd = f'GIT_TERMINAL_PROMPT=0 git clone --depth=1 --single-branch {branch_cmd} "{remote_url}" "{workspace.resolve()}"'
+            shell(cmd, workdir=project_root(), print_log=False)
         except CommandError as e:
             if 'fatal: Authentication failed' in e.stdout:
                 if git_credentials is None:
                     raise RuntimeError('no git credentials configured for a repo') from e
                 raise ContextError('authentication failure') from e
             raise e
 
-    assert workspace.is_dir(), f'workspace directory doesn\'t exist: {workspace}'
-
-    repo = Repo(workspace.resolve())
-    with wrap_context('checking out git branch'):
-        if manifest.git.branch:
-            repo.git.checkout(manifest.git.branch)
+    git_version = read_job_git_version(workspace)
+    shutil.rmtree((workspace / '.git').resolve())  # exclude .git from the job image
```

Comment:
This makes me nervous. Maybe add an extra check to make sure `workspace` is properly defined? I don't think it's likely to happen, nor likely to do a whole lot of damage, but better safe than sorry when it comes to recursive removal.

Reply:
Something like this
```
    git_path = (workspace / '.git').resolve()
    assert git_path.as_posix().startswith(str(project_root() / 'workspaces'))
```
 will make you reassured?